### PR TITLE
Allow InvokeMethodRecursively to take null arguments

### LINKED ISSUE #25: ObjectMethodInvokeExtensions.Invoke*MethodRecursively fails with null arguments
Issue moved from rapidcore/issues#2

The current implementation of `ObjectMethodInvokeExtensions.Invoke*MethodRecursively` fails if you supply `null` as an argument for the method that should be called.

This is due to the underlying call to the `GetMethodRecursively`, which is using an array of the types from the arguments to find the method to call. This however does not support the null arguments as it does not have logic to "guess" the type of the corresponding argument.

The `bugfix/objectmethodinvokeextensions-null-handling` branch contains a commit with tests that show the issue and a commit with some work-in-progress code for an attempted solution (which is basically to use the null arguments as wildcards while going through all methods with the given name.

### PR DESCRIPTION
**Pull request by IceAndSnow**
Closes #25

### CODE REVIEW

**Review by nover — COMMENTED** [review]
I think it looks OK but I will defer the merge to @pingvinen

File: src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs [hunk 1]
```diff
@@ -17,11 +17,50 @@ public static class TypeGetMethodRecursivelyExtensions
...
+                .ToList()
+                .FindAll(m =>
+                {
+                    // Make sure that each parameter satisfies the method definition
+                    for (var parameterIndex = 0; parameterIndex < argTypes.Length; parameterIndex++)
+                    {
+                        if (argTypes[parameterIndex] == null)
+                        {
+                            var parameter = m.GetParameters()[parameterIndex].ParameterType;
+                            if (parameter.GetTypeInfo().IsValueType && Nullable.GetUnderlyingType(parameter) == null)
```
Comment (pingvinen) [inline src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs]: Could this be replaced by `if (parameter.IsNullable())`

File: src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs [hunk 2]
```diff
@@ -17,11 +17,50 @@ public static class TypeGetMethodRecursivelyExtensions
...
+                            {
+                                // If the argument given is null but the parameter is not nullable
+                                return false;
+                            }
+                        }
+                        else
+                        {
+                            if (m.GetParameters()[parameterIndex].ParameterType != argTypes[parameterIndex])
```
Comment (pingvinen) [inline src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs]: I wonder if we actually need to check this slightly differently so that we allow people to provide `5` for a parameter that is actually of type `int?`.

E.g.

```csharp
if (m.GetParameters()[parameterIndex].ParameterType.GetTypeOrUnderlyingNullableType() != argTypes[parameterIndex])
```

**Review by pingvinen — CHANGES_REQUESTED** [review]
It looks good, but I think we need some tests that prove that we are handling method signatures with nullables correctly.

File: src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs (hunk 2 above)
Comment (IceAndSnow) [reply src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs]: I agree that this needs to happen a bit differently. I created an issue for this problem: #110  However perhaps this should simply be fixed here.

File: src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs (hunk 1 above)
Comment (IceAndSnow) [reply src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs]: I don't believe this is the case as `parameter.IsNullable()` returns false for the string type.

File: src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs (hunk 2 above)
Comment (IceAndSnow) [reply src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs]: @pingvinen But is it actually desirable to do it in this method?
Think of this case:

```
public class SuperAwesomeBanana
{
    public void DoSomethingAwesome(int i) { /*Actually do something awesome*/ }
    public void DoSomethingAwesome(int? i) { /*Do something different but awesome*/ }
}
```

Although this probably is not a good code practice, this would create some problems if the underlying type is always considered, as these two methods would both satisfy a call to `someType.GetMethodRecursively("DoSomethingAwesome", typeof(int))`.

Perhaps an entirely new method should be created: `GetMethodRecursivelyIgnoringNullable` that does exactly the same as the normal `GetMethodRecursively` however it treats `int` and `int?` the same.

File: src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs (hunk 1 above)
Comment (pingvinen) [reply src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs]: Ah yes, because string is this weird value type that can be set to null, but is not nullable.

Either we could "wrap" this in an extension method on `Type` like `Type.CanBeSetToNull`. This would make the code more semantic and thus easier to read.

File: src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs (hunk 2 above)
Comment (pingvinen) [reply src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs]: Nice example :) I would argue that it is a code smell that should be fixed in the consumer's codebase, but it is still a case that we are now aware of and is easy to handle.

If we call our method with `5` then we should find both methods and therefore throw an `AmbiguousMatchException`. If we call our method with `null` then we will only find the nullable method.

Anywho... add a unit test that shows this behavior.

Comment (IceAndSnow) [reply src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs]: @pingvinen Consider the `SuperAwesomeBanana` example again. What about a call like this:

```
typeof(SuperAwesomeBanana).GetMethodRecursively("DoSomethingAwesome", typeof(int?))
```

I find this a bit more tricky. I would say that the justification that `GetMethodRecursively` finds both overloaded methods when supplying the type: `typeof(int)`, is that it finds every method with the given name that always accepts parameters of the given types. However I would argue that the above mentioned call, where `typeof(int?)` is supplied should succeed by only finding one of the overloaded methods. This is because not all `int?`-values are applicable for both overloaded methods.

In other words my proposal is to change the code to the following:

```
var parameter = m.GetParameters()[parameterIndex].ParameterType;
if (parameter != argTypes[parameterIndex] &&
    parameter.GetTypeOrUnderlyingNullableType() != argTypes[parameterIndex])
{
    // If the argument and parameter does not match
    return false;
}
```

Comment (pingvinen) [reply src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs]: > This is because not all int?-values are applicable for both overloaded methods.

I am not sure I follow...

I think that a direct call to `GetMethodRecursively` should return _the_ method matching the given signature. The compiler does not allow multiple methods with the same signature, so that should be fairly straightforward. And remember that `int != int?`.

The issue we are actually trying to solve is that if `InvokeMethodRecursively` is called with a `null` input, it has no way of telling the underlying call to `GetMethodRecursively` what the type of that argument is.

E.g.

```csharp
class DinosaursHadLegs
{
    public string DoMagic(string one, string two)
    {
    }

    public string DoMagic(string one, int two)
    {
    }

    public string DoMagic(string one, int? two)
    {
    }

    public string DoMagic(string one, NoTheyDidNot two)
    {
    }
}


var instance = new DinosaursHadLegs();

instance.InvokeMethodRecursively(nameof(DinosaursHadLegs.DoMagic), "one", "two");
// should have called DoMagic(string one, string two)

instance.InvokeMethodRecursively(nameof(DinosaursHadLegs.DoMagic), "one", 5);
// multiple matching signatures
// DoMagic(string one, int two)
// DoMagic(string one, int? two)


instance.InvokeMethodRecursively(nameof(DinosaursHadLegs.DoMagic), "one", null);
// multiple matching signatures
// DoMagic(string one, string two)
// DoMagic(string one, int? two)
// DoMagic(string one, NoTheyDidNot two)
```

I think `GetMethodRecursively` should treat `null` as `not sure about the type, but it must support being set to null`.

**Review by pingvinen — APPROVED** [review]
